File: crc32/encoder.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// Convierte una cadena a su representación binaria
func stringToBinary(s string) string {
	var result string
	for _, c := range s {
		result += fmt.Sprintf("%08b", c)
	}
	return result
}

// Convierte un entero a su representación binaria
func intToBinary(n int) string {
	return fmt.Sprintf("%064b", n)
}

// Función genérica para convertir diferentes tipos de datos a binario
func toBinary(data string) (string, error) {
	if n, err := strconv.Atoi(data); err == nil {
		return intToBinary(n), nil
	} else {
		return stringToBinary(data), nil
	}
}

// Cálculo manual del CRC32
func crc32(data []byte) uint32 {
	crc := ^uint32(0)
	for _, b := range data {
		crc ^= uint32(b)
		for i := 0; i < 8; i++ {
			if crc&1 == 1 {
				crc = (crc >> 1) ^ 0xEDB88320
			} else {
				crc >>= 1
			}
		}
	}
	return ^crc
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Ingrese el mensaje: ")
	input, _ := reader.ReadString('\n')
	input = strings.TrimSpace(input)

	binaryData, err := toBinary(input)
	if err != nil {
		log.Fatalf("Error converting to binary: %v", err)
	}

	// Cálculo del CRC32
	crc := crc32([]byte(input))
	crcBinary := fmt.Sprintf("%032b", crc)

	encodedMessage := binaryData + crcBinary

	fmt.Println("Encoded message:", encodedMessage)
}

File: crc32/encoder.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <iomanip>

// Función para calcular el CRC32
uint32_t crc32(const std::string& data) {
    uint32_t crc = 0xFFFFFFFF;
    for (char c : data) {
        crc ^= static_cast<uint8_t>(c);
        for (int i = 0; i < 8; i++) {
            if (crc & 1)
                crc = (crc >> 1) ^ 0xEDB88320;
            else
                crc >>= 1;
        }
    }
    return crc ^ 0xFFFFFFFF;
}

std::string to_hex_string(uint32_t num) {
    std::stringstream ss;
    ss << std::hex << std::setw(8) << std::setfill('0') << num;
    return ss.str();
}

int main() {
    std::string input;
    std::cout << "Ingrese el mensaje en binario: ";
    std::cin >> input;

    uint32_t crc = crc32(input);
    std::string crc_hex = to_hex_string(crc);

    std::cout << "Mensaje codificado: " << input << crc_hex << std::endl;
    return 0;
}
